full-moon/src/atom.rs

```diff
@@ -0,0 +1,291 @@
+use logos::{Lexer, Logos};
+
+use crate::ShortString;
+
+pub(crate) fn trim_bracket_head(slice: &str) -> (ShortString, Option<usize>) {
+    match test_bracket_head(slice) {
+        Some(count) => {
+            let trim = &slice[count + 2..slice.len() - count - 2];
+
+            (trim.into(), Some(count))
+        }
+        None => (slice.into(), None),
+    }
+}
+
+fn test_bracket_head(slice: &str) -> Option<usize> {
+    // starts with `[`?
+    if !slice.starts_with('[') {
+        return None;
+    }
+
+    // how many `=` after `[`?
+    let count = slice.chars().skip(1).take_while(|&v| v == '=').count();
+
+    // ends with `[`?
+    if !matches!(slice.chars().nth(count + 1), Some('[')) {
+        return None;
+    }
+
+    Some(count)
+}
+
+fn read_bracketed(lex: &mut Lexer<Atom>, skips: usize) -> bool {
+    let num_eq = match lex.slice().get(skips..).and_then(test_bracket_head) {
+        Some(v) => v,
+        None => return false,
+    };
+
+    let mut search = false;
+    let mut num = 0;
+
+    for (i, v) in lex.remainder().char_indices() {
+        match (search, v) {
+            (true, '=') => num += 1,
+            (true, ']') if num_eq == num => {
+                lex.bump(i + 1);
+
+                return true;
+            }
+            (false, ']') => {
+                search = true;
+                num = 0;
+            }
+            _ => search = false,
+        }
+    }
+
+    false
+}
+
+#[derive(Logos, Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) enum Atom {
+    #[token("and")]
+    And,
+
+    #[token("break")]
+    Break,
+
+    #[token("do")]
+    Do,
+
+    #[token("else")]
+    Else,
+
+    #[token("elseif")]
+    ElseIf,
+
+    #[token("end")]
+    End,
+
+    #[token("false")]
+    False,
+
+    #[token("for")]
+    For,
+
+    #[token("function")]
+    Function,
+
+    #[token("if")]
+    If,
+
+    #[token("in")]
+    In,
+
+    #[token("local")]
+    Local,
+
+    #[token("nil")]
+    Nil,
+
+    #[token("not")]
+    Not,
+
+    #[token("or")]
+    Or,
+
+    #[token("repeat")]
+    Repeat,
+
+    #[token("return")]
+    Return,
+
+    #[token("then")]
+    Then,
+
+    #[token("true")]
+    True,
+
+    #[token("until")]
+    Until,
+
+    #[token("while")]
+    While,
+
+    #[cfg(feature = "lua52")]
+    #[token("goto")]
+    Goto,
+
+    #[token("+=")]
```

Comment:
These are all Luau specific, and should be behind the feature flag.

If they weren't before, I forget the reason for it, and I guess we'll see.